### PR TITLE
Feat: Two ravel implementations of naive all-sky imaging

### LINKED ISSUE #21: D2.2 Initial baseline benchmarks
Establish and document reference performance by running the agreed benchmark suite on consistent baseline implementations and environments to provide the starting point for all subsequent optimizations.

- Baseline setup: Finalize and document the hardware/software environment defined in [D2.1](https://github.com/astron-rd/PACE/issues/20)
- Reference implementations: Provide minimally optimized Python and/or C++ versions of each target module as the baseline.
- Execution of benchmarks: Run the benchmarks defined in D2.1 across all reference implementations.
- Data collection: Capture raw measurements following the reporting procedures from D2.1.
- Analysis: Summarize baseline performance, identify bottlenecks, and establish reference values for later optimization.
- Documentation: Present results in clear tables/plots with a concise summary of initial findings.

### PR DESCRIPTION
**Pull request by Dantali0n**
This add two implementations of the most basic all-sky imaging algorithm (mapping an image plane onto a celestial sphere) this algorithm uses antenna pair baselines (UVW) combined with correlated Visibilities (Sometimes called XSTs in the context of LOFAR) to produce a 2D image.

The algorithm closely relates to a 2D FFT and is often parallelizeable with good scaling and speedup.

in this basic implementation we add two versions one using the popular library Numba and the other using the less well known library Jax.

In earlier trials we have observed that Jax typically achieves orders of magnitude better performance compared to Numba hence our reasoning for adding these.

The goal is to also add Jax implementations for the IDG Python kernels.

I will fix the packaging of this sub project and adding the linting to pre-commit once #42 is merged.

Below is a table with some rudimentary benchmarks, the environment wasn't noise free so take them with some salt.

| Device | Algorithm           | JIT | X   | Y   | Min   | Max    | Mean   | std dev | Joules / It |
|--------|---------------------|-----|-----|-----|-------|--------|--------|---------|-------------|
| 155H   | Numpy Ravel Real  | N.A | 256 | 256 | 6.29 | 6.51 | 6.39   | 0.07    |  187.18    |
| 155H   | Numba Ravel Real    | Yes | 256 | 256 | 2.65  | 2.83   | 2.71   | 0.04    |  149.49   |
| 155H   | Jax Ravel Real      | Yes | 256 | 256 | 0.51  | 0.56   | 0.53   | 0.01    | 22.48       |
| 1050 TI | Jax Ravel Real | Yes | 256 | 256 | 0.058  | 0.0788 | 0.073  | 0.006   |
| 7900 XTX | Jax Ravel Real | Yes | 256 | 256 | 0.0059 | 0.0333 | 0.0073 | 0.0049  | 1.38  |

### CODE REVIEW

File: imaging/python/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[project]
+name = "imaging-python"
+version = "0.1.0"
+requires-python = "==3.10"
```
Comment (csbnw) [inline imaging/python/pyproject.toml]: Why Python 3.10?

I ran this on a Debian 13 machine:
```
module load spack/20251001-deb13
module load python py-pip python-venv
python3 -m venv env
source env/bin/activate
pip install pip-compile
pip-compile pyproject.toml
pip-sync
```
And got Python 3.12.9, which seems to work fine:
```
$ time python3 imaging_numba.py 

real    0m0.248s
user    0m1.427s
sys     0m4.074s

$ time python3 imaging_jax.py 

real    0m0.580s
user    0m1.380s
sys     0m4.227s
```

Comment (Dantali0n) [reply imaging/python/pyproject.toml]: `nptyping` was incompatible with `numpy>=2` which required Python 3.11 or higher for a long time. They have updated nptyping if you can use more recent versions

**Comment by Dantali0n** [comment]
> Finally, as someone unfamiliar with Jax (other than what you told me about it), I don't know how to run the code using a GPU. This is a great opportunity to (briefly) describe how it works on the [wiki](https://github.com/astron-rd/PACE/wiki/PACE).

I'd rather not regurgitate existing documentation of external projects as A; it has a tendency to go out of date, B; has already been written down by someone else much more familiar with the topic then I am. But I would be okay with creating a Wiki page with several references and pointers on the subject that link to the official documentation.

**Comment by Dantali0n** [comment]
### Lofty IDG?
I don't fully understand this question, could you elaborate? "Should we Lofty for the IDG Python code too?"

### Timing
All the timing and measuring problems indeed are easily overcome, we can just use `timeit`.
There is just one small detail to think about, I will draw it out here:

**Some Jax gotcha we need to deal with:**
Jax uses asynchronous dispatch when running on accelerated devices like GPU / NPU. When reading the results it is lazily loaded once it becomes available. But the compute call returns immediately. This means it can't be used for timing measurements.

Jax asynchronous dispatch calls can use the `block_until_ready()` method to await completion. However, none async dispatch calls do not have this method.

The result is that for generalized measurements you will need some wrapping.

```Python
if has_attr(obj, 'block_until_ready'):
    obj.block_until_ready()
```

Now suppose the CPU is really really slow, and both the dispatch and kernel have long completed before the `if` condition is evaluated, the results of `timeit` will also absorb the runtime cost of the very slow CPU.

Until now, I have not dealt with this but I think we should, so I propose that for measuring Jax kernels we will create some logic based on the chosen Jax runtime (CPU, GPU, NPU, etc) Where we create two methods and dynamically switch between the one that is measured. That way one method can always calll `block_until_ready` and the other never does without the cost of the `if` branch in either definitions.

### Wiki

I will work on it and then we can maybe merge? I will update this post with the link once available and reference it in the codebase

**Comment by csbnw** [comment]
Could we do a warm-up run first and take a timing measurement for a second run?

How about using [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/usage.html)?

Extending https://github.com/astron-rd/PACE/blob/main/idg/python/main.py#L185, something like this could work:

```
class JAXTimer:
    def __init__(self, description: str, warmup: int = 1):
        self.description = description
        self.warmup = warmup
        self.duration = None
    
    def __enter__(self):
        return self
    
    def __exit__(self, *args):
        pass
    
    def run(self, operation: Callable, *args, **kwargs):
        # warmup
        for _ in range(self.warmup):
            result = operation(*args, **kwargs)
            jax.block_until_ready(result)
        
        # timing
        start = time.perf_counter()
        result = operation(*args, **kwargs)
        jax.block_until_ready(result)
        self.duration = time.perf_counter() - start
        
        print(f"{self.description:<38} {self.duration:>9.6f} s")
        _timings[self.description] = self.duration
        return result
        
    def __call__(self, operation: Callable):
        return self.run(operation)
```

Or even fancier, add a decorator?

**Review by mickveldhuis — COMMENTED** [review]
Cool stuff! Just mainly left some comments to improving the readability 😁 It's always a little tricky when you decompose a problem further and further to optimise.

File: all-sky/python/all_sky_numba.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    p, q, r = M.shape  # 96, 96, 51040
+    out = np.empty((p, q, r), dtype=M.dtype)
+    for i in prange(p):
+        for j in range(q):
+            for k in range(r):
+                out[i, j, k] = (s * M[i, j, k])[0]
+    return out
+
+
+# Do not jit, worse performance
```
[the remark this answers is not in the document]
Comment (Dantali0n) [reply all-sky/python/all_sky_numba.py]: 23% or so, but I fiddled with the numba implementation some more and its now slightly better, will update the patch

**Comment by Dantali0n** [comment]
Hello @csbnw and @mickveldhuis I have addressed your review comments hoping we can round this off. Note, I also added a cheeky extra config option to the Jax config that makes it even faster on CPU.

**Review by mickveldhuis — APPROVED** [review]
@Dantali0n I'm happy to approve this PR.
